### PR TITLE
 Fix 2399

### PR DESCRIPTION
**Pull request by marharyta**
# Definition of Done / Review checklist

## Reviewability
- [x] link to issue - https://github.com/CSCfi/rems/issues/2399
- [ ] note if PR is on top of other PR 
- [ ] note if related change in rems-deploy repo
- [ ] consider adding screenshots for ease of review

## API
- [ ] API is documented and shows up in Swagger UI
- [ ] API is backwards compatible or completely new
- [ ] Events are backwards compatible

## Documentation
- [ ] update changelog if necessary
- [ ] add or update docstrings for namespaces and functions
- [ ] components are added to guide page
- [ ] documentation _at least_ for config options (i.e. docs folder)
- [ ] ADR for major architectural decisions or experiments

## Different installations
- [ ] new configuration options added to rems-deploy repository
- [ ] instance specific translations (i.e. LBR kielivara)

## Testing
- [ ] complex logic is unit tested
- [ ] valuable features are integration / browser / acceptance tested automatically

## Accessibility
- [ ] all icons have the aria-label attribute
- [ ] all fields have a label
- [ ] errors are linked to fields with aria-describedby
- [ ] contrast is checked
- [ ] conscious decision about where to move focus after an action

## Follow-up
- [ ] new tasks are created for pending or remaining tasks
- [ ] no critical TODOs left to implement

### CODE REVIEW

**Comment by marharyta** [comment]
There is a TODO still:
- [ ] add proper translation to "unsupported text" I believe @Macroz  has had a couple of suggestions
- [ ] I am not sure anymore weather we still have anything else to adhere to the plan A: "alternative A (the one I have been talking about):
input components: not shown for unsupported fields
app db: can still contain unsupported fields
validation: based on result of build-request
pros: the actual request that would be sent to the server is validated. no need to clear app-db. no data is lost if switching back and forth between field types
cons: build-request code needs to be in sync with input component showing code, but that's accomplished with common-form/supports-info-text? etc."

File: src/cljc/rems/common/form.cljc [hunk 1]
```diff
@@ -239,6 +246,25 @@
       (assoc form :form/fields fields)
       form)))
 
+;; yaya, this one fails
```
Comment (opqdonut) [inline src/cljc/rems/common/form.cljc:249]: ?

File: src/cljc/rems/common/form.cljc [hunk 2]
```diff
@@ -239,6 +246,25 @@
...
+(deftest validate-fields-compatibility-test
```
Comment (opqdonut) [inline src/cljc/rems/common/form.cljc:250]: Tests, especially tests like this that are in source files, are usually named like `test-validate-fields-compatiblity`. That way they don't interfere with e.g. function name completion.

File: src/cljc/rems/common/form.cljc [hunk 3]
```diff
@@ -183,7 +183,7 @@
 
 (defn- validate-not-present [field key]
   (when (contains? field key)
-    {key :unsupported}))
+    {key {key :t.form.validation/unsupported}}))
```
Comment (opqdonut) [inline src/cljc/rems/common/form.cljc:186]: I think you can revert this change now. The frontend shouldn't see any `:unsupported` validations any more.

File: src/cljc/rems/common/form.cljc [hunk 4]
```diff
@@ -224,6 +226,11 @@
       remove-empty-keys
       nil-if-empty))
 
+(defn validate-fields-compatibility [form languages]
+  (-> {:form/fields (validate-fields (:form/fields form) languages)}
+      remove-empty-keys
+      nil-if-empty))
```
Comment (opqdonut) [inline src/cljc/rems/common/form.cljc:232]: is this function unused? if you want to keep it, a docstring would be nice

File: src/cljc/rems/common/form.cljc [hunk 5]
```diff
@@ -210,6 +210,8 @@
                           (validate-not-present field :field/visibility))))})]
     (apply merge (map-indexed validate-field fields))))
 
+
+
```
Comment (opqdonut) [inline src/cljc/rems/common/form.cljc:214]: note some spurious changes here

File: src/cljs/rems/administration/create_form.cljs [hunk 6]
```diff
@@ -177,13 +177,14 @@
  ::send-form
  (fn [{:keys [db]} [_]]
    (let [edit? (::edit-form? db)
-         form-errors (validate-form-template (get-in db [::form :data]) (:languages db))
+         form-preview-data (build-request (get-in db [::form :data]) (:languages db))
+         form-errors (validate-form-template form-preview-data (:languages db))
```
Comment (opqdonut) [inline src/cljs/rems/administration/create_form.cljs:181]: huh, that's all the fix really needs? cool

File: src/cljs/rems/administration/create_form.cljs [hunk 7]
```diff
@@ -177,13 +177,14 @@
  ::send-form
  (fn [{:keys [db]} [_]]
    (let [edit? (::edit-form? db)
-         form-errors (validate-form-template (get-in db [::form :data]) (:languages db))
+         form-preview-data (build-request (get-in db [::form :data]) (:languages db))
```
Comment (opqdonut) [inline src/cljs/rems/administration/create_form.cljs:180]: how about calling this `request` instead of `form-preview-data`?

**Review by opqdonut — CHANGES_REQUESTED** [review]
some comments / questions

looks good in general

File: src/cljc/rems/common/form.cljc (hunk 1 above)
Comment (marharyta) [reply src/cljc/rems/common/form.cljc:249]: oops

File: src/cljs/rems/administration/create_form.cljs (hunk 7 above)
Comment (marharyta) [reply src/cljs/rems/administration/create_form.cljs:180]: You mean re-naming or?

File: src/cljs/rems/administration/create_form.cljs (hunk 6 above)
Comment (marharyta) [reply src/cljs/rems/administration/create_form.cljs:181]: well, it works (on my machine haha) and acts according to the spec

File: src/cljs/rems/administration/create_form.cljs (hunk 7 above)
Comment (opqdonut) [reply src/cljs/rems/administration/create_form.cljs:180]: Yeah, renaming the local. Just a small suggestion, not insisting

File: src/cljc/rems/common/form.cljc (hunk 3 above)
Comment (Macroz) [reply src/cljc/rems/common/form.cljc:186]: And I guess also the text localizations can be removed?

Comment (opqdonut) [reply src/cljc/rems/common/form.cljc:186]: yeah that's ok too, or we can keep the translation key & translations for "defense in depth"

File: src/cljc/rems/common/form.cljc (hunk 5 above)
Comment (marharyta) [reply src/cljc/rems/common/form.cljc]: yes, should be fixed by now though

File: src/cljc/rems/common/form.cljc (hunk 4 above)
Comment (marharyta) [reply src/cljc/rems/common/form.cljc]: Should be fixed by now

File: src/cljc/rems/common/form.cljc (hunk 2 above)
Comment (marharyta) [reply src/cljc/rems/common/form.cljc]: It is deleted now

File: src/cljc/rems/common/form.cljc (hunk 3 above)
Comment (marharyta) [reply src/cljc/rems/common/form.cljc:186]: I think we should keep it

Comment (opqdonut) [reply src/cljc/rems/common/form.cljc:186]: so then just revert to `{key :t.form.validation/unsupported}`

File: src/cljc/rems/common/form.cljc [hunk 8]
```diff
@@ -291,14 +301,15 @@
              (validate-form-template (assoc-in form [:form/fields 0 :field/info-text] {:en "en info text" :fi ""}) languages)
              (validate-form-template (assoc-in form [:form/fields 0 :field/info-text] {:en "en info text"}) languages))))
 
+    ;; fix here
```
Comment (opqdonut) [inline src/cljc/rems/common/form.cljc:304]: leftover comment?

**Review by opqdonut — APPROVED** [review]
looks good, remove that one comment though
